smithy-protocol-test-traits/src/main/java/software/amazon/smithy/protocoltests/traits/HttpRequestTestCase.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.protocoltests.traits;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import software.amazon.smithy.model.node.ArrayNode;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+/**
+ * Defines a test case for an HTTP request.
+ */
+public final class HttpRequestTestCase extends HttpMessageTestCase implements ToSmithyBuilder<HttpRequestTestCase> {
+
+    private static final String METHOD = "method";
+    private static final String URI = "uri";
+    private static final String QUERY_PARAMS = "queryParams";
+    private static final String FORBID_QUERY_PARAMS = "forbidQueryParams";
+    private static final String REQUIRE_QUERY_PARAMS = "requireQueryParams";
+
+    private final String method;
+    private final String uri;
+    private final Map<String, String> queryParams;
+    private final List<String> forbidQueryParams;
+    private final List<String> requireQueryParams;
+
+    private HttpRequestTestCase(Builder builder) {
+        super(builder);
+        method = SmithyBuilder.requiredState(METHOD, builder.method);
+        uri = SmithyBuilder.requiredState(URI, builder.uri);
+        queryParams = Collections.unmodifiableMap(new LinkedHashMap<>(builder.queryParams));
```

Comment:
We have `MapUtils.copyOf` if you don't need to go through a `LinkedHashMap` specifically.

Reply:
Keeping the order might be useful. Let's keep it this way for now.